pkg/resource/policy/hooks.go

```diff
@@ -188,6 +219,246 @@ func (rm *resourceManager) removeTags(
 	input.TagKeys = inTagKeys
 
 	_, err = rm.sdkapi.UntagPolicyWithContext(ctx, input)
-	rm.metrics.RecordAPICall("DELETE", "UntagPolicy", err)
+	rm.metrics.RecordAPICall("UPDATE", "UntagPolicy", err)
+	return err
+}
+
+// updatePolicyDocument creates a new Policy version with the new
+// PolicyDocument and returns the newly-created version ID.
+//
+// A policy is technically immutable. In order to modify the PolicyDocument,
+// one calls the CreatePolicyVersion API call and creates a new Policy version
+// with the updated PolicyDocument.
+func (rm *resourceManager) updatePolicyDocument(
+	ctx context.Context,
+	r *resource,
+) (string, error) {
+	var err error
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.updatePolicyDocument")
+	defer func(err error) {
+		exit(err)
+	}(err)
+
+	policyARN := (*string)(r.ko.Status.ACKResourceMetadata.ARN)
+
+	if err = rm.ensureVersionsLimitNotExceeded(ctx, *policyARN); err != nil {
+		return "", err
+	}
+
+	input := &svcsdk.CreatePolicyVersionInput{}
+	input.PolicyArn = policyARN
+	input.PolicyDocument = r.ko.Spec.PolicyDocument
+
+	trueVal := true
+	input.SetAsDefault = &trueVal
+
+	resp, err := rm.sdkapi.CreatePolicyVersionWithContext(ctx, input)
+	rm.metrics.RecordAPICall("UPDATE", "CreatePolicyVersion", err)
+	if err != nil {
+		return "", err
+	}
+	return *resp.PolicyVersion.VersionId, err
+}
+
+// ensureVersionsLimitNotExceeded checks to see if the number of versions
+// for a supplied managed policy ARN exceeds 4 and deletes the oldest policy
+// version if so.
+//
+// According to the IAM docs:
+//
+// > A managed policy can have up to five versions. If the policy has five
+// > versions, you must delete an existing version using DeletePolicyVersion
+// > before you create a new version.
+//
+// https://docs.aws.amazon.com/IAM/latest/APIReference/API_CreatePolicyVersion.html
+func (rm *resourceManager) ensureVersionsLimitNotExceeded(
+	ctx context.Context,
+	policyARN string,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.ensureVersionsLimitNotExceeded")
+	defer func(err error) {
+		exit(err)
+	}(err)
+
+	versions, err := rm.getPolicyVersions(ctx, policyARN)
+	if err != nil {
+		return err
+	}
+
+	if len(versions) == limitPolicyVersions {
+		err = rm.deletePolicyVersion(ctx, policyARN, versions[0].version)
```

Comment:
is there a need to check that versions[0].version is not default version?

Reply:
I suppose it wouldn't hurt, no...

Reply:
done.